WebSocket refactoring, bug fixes, improvements, and passes most Autobahn Test (The most standard WebSocket Protocol Test).

dumganhar opened the pull request:
This PR is based on the latest release version of `libwebsockets`, so before merging it, please upgrade `libwebsockets`
## Result

It has passed most of Autobahn Test (The most standard WebSocket Protocol Test).
Please see the result at http://182.254.241.97/runtime-test/websocket-report/index.html
## Changes
### Refactoring
- Adds more callback functions like `onConnectionOpened`, `onConnectionClosed` rather than put all websocket callback in just one `onSocketCallback`, code will be clearer and readable.

``` c++
    int onSocketCallback(struct lws *wsi, int reason, void *user, void *in, ssize_t len);

    void onClientWritable();
    void onClientReceivedData(void* in, ssize_t len);
    void onConnectionOpened();
    void onConnectionError();
    void onConnectionClosed();
```
- Uses `performFunctionInCocosThread(cb);` to send msg to Cocos Thread rather than make a scheduler update
- Uses `LOGD` for outputting log to debugging websocket connections since websocket stuff uses a sub thread. CCLOG isn't threadsafe.
### Bug Fixes
- WebSocket::close should not block Cocos Thread, it should just notify websocket thread to exit. Now the behavior is the same as Browser. To prove, you could type some test code in Chrome and the result will be:

``` javascript
ws = new WebSocket("ws://127.0.0.1:9123"); ws.onclose = function(){ console.log("I'm closed");}; 
(){ console.log("I'm closed");}
console.log("before close"); ws.close(); console.log("hello");
VM437:2 before close
VM437:2 hello
undefined
VM433:2 I'm closed
```
- Fixed every websocket event will cost 50ms at least, since the old way was wrong, we should not wait 50ms in sub thread while loop. `lws_service(xxx, time_to_wait)` will do it for us. So it fixed delay of message receive/send.
- Fixed potential high CPU cost since the old code will trigger a websocket writable event every time after the writable callback. Now only when the size of the sending message queue isn't zero, a `lws_callback_on_writable` will be invoked
- Fixed `libwebsocket_context_destroy` may not be invoked while closing connection.
- Fixed sending large data will trigger websocket errors
- Fixed receiving large data in many frames will trigger JS `onmessage` callback many times
- Fixed `lws_write` may return a smaller value than the value `n` for sending, a `WebSocketFrame` class is doing the job. `data->issued` should not be increased while a `WebSocketFrame` isn't totally sent.
- Fixed connection isn't closed if received text message is an invalid UTF8 string.
- Fixed the memory leak that JSB_WebSocketDelegate is never released.
### Improvements
- Better performance for received data buffer. Uses `std::vector` for dynamic received data buffer, the old way is every time there is a frame received, buffer memory will be extended every time there is a message comes. It was very low performance.
- Supports receive/send data with zero size.
- Larger socket rx/tx buffer size `WS_RX_BUFFER_SIZE` for better performance.
- Adds static method `WebSocket::closeAllConnections`, it should be invoked while resetting `Director`, for ensure all websocket threads exit. Otherwise, cocos will release all engine classes when exiting game, then if a websocket event is coming, it will cause exception since websocket API is asynchronous.

Commit:
WebSocket refactoring, bug fixes, improvements, and passes most Autobahn Test (The most standard WebSocket Protocol Test).

Commit:
Adds WebSocket::closeAllConnections() method.

Commit:
Invoking WebSocket::closeAllConnection while reseting Director.

pandamicro commented on cocos/network/WebSocket.cpp:
```diff
@@ -120,27 +170,22 @@ WsThreadHelper::WsThreadHelper()
+
     // Creates websocket thread
-    _subThreadInstance = new (std::nothrow) std::thread(&WsThreadHelper::wsThreadEntryFunc, this);
+    _subThreadInstance = new std::thread(&WsThreadHelper::wsThreadEntryFunc, this);
```
Why remove nothrow declaration ?

pandamicro commented on cocos/network/WebSocket.cpp:
```diff
@@ -151,22 +196,22 @@ void WsThreadHelper::quitSubThread()
 }
 
-void WsThreadHelper::sendMessageToUIThread(WsMessage *msg)
+void WsThreadHelper::sendMessageToUIThread(const std::function<void()>& cb)
```
This function name seems no longer suitable

pandamicro commented on cocos/network/WebSocket.cpp:
```diff
@@ -241,21 +323,42 @@ WebSocket::WebSocket()
+        }
+        else
+        {
+            LOGD("ERROR: WebSocket instance (%p) added to container!\n", this);
```
To be more clear
ERROR: WebSocket instance (%p) haven't been added to global list!

pandamicro commented on cocos/network/WebSocket.cpp:
```diff
@@ -311,33 +414,35 @@ bool WebSocket::init(const Delegate& delegate,
-	_wsProtocols = new (std::nothrow) libwebsocket_protocols[protocolCount+1];
-	memset(_wsProtocols, 0, sizeof(libwebsocket_protocols)*(protocolCount+1));
+
+    _wsProtocols = new lws_protocols[protocolCount+1];
```
Normally, we should always add `nothrow` to new statement, riq and Wenhai have cleaned up our codes, if there is no special reason, we shouldn't remove it.

pandamicro commented on cocos/network/WebSocket.cpp:
```diff
@@ -311,33 +414,35 @@ bool WebSocket::init(const Delegate& delegate,
         for (std::vector<std::string>::const_iterator iter = protocols->begin(); iter != protocols->end(); ++iter, ++i)
         {
-            char* name = new (std::nothrow) char[(*iter).length()+1];
+            char* name = new char[(*iter).length()+1];
```
Same as above

pandamicro commented:
Thanks @dumganhar Great PR !

The logic is much more clear with better stability and performance. I have no further questions except some minor remark in the previous comments. However, I'm not very familiar with network behavior or websocket, and I'm not the owner of network module. @linshun Can you assign someone to give a final confirm on this PR ?

dumganhar commented:
@pandamicro ,all you commented were fixed.

Commit:
Comment fix & function name fix for WsThreadHelper:
sendMessageToUIThread -> sendMessageToCocosThread
sendMessageToSubThread -> sendMessageToWebSocketThread
joinSubThread -> joinWebSocketThread

linshun commented:
Hi @ricardoquesada, you are the owner of networking and could you please review this PR and merge it? Thanks.

It is a great improvement for WebSocket.

ricardoquesada commented:
looks good to me.

@dumganhar what is the use case for cocos2d-x ? Is it only valid for cocos2d-js users ?
BTW, the build is red. If you fix it, I'll merge it.

Commit:
Updates WebSocket::onSocketCallback & WebSocketTest.js

Commit:
LWS_SEND_BUFFER_PRE_PADDING -> LWS_PRE

Commit:
Removes lws_get_internal_extensions, adds permessage-deflate extension.

Commit:
More function name renames. createThread -> createWebSocket, quitSubThread -> quitWebSocketThread.

dumganhar commented:
@ricardoquesada , i'll update libwebsocket tomorrow.

Commit:
whitespace cleanup.

dumganhar commented:
Updates https://github.com/cocos2d/cocos2d-x-3rd-party-libs-src/pull/78
Could somebody build the libwebsockets library for all platforms?

Commit:
Don't change the behavior of WebSocket::close method. WebSocket::close is an synchronous method before, adds a WebSocket::closeAsync method for asynchronous closing.

Commit:
_readyState must be set to State::CLOSED and _delegate->onClose(this) must be invoked at the end of WebSocket::close.

Commit:
'closed' state has to be set before quit websocket thread.

Commit:
Adds isDestroy std::share_ptr variable to dectect whether websocket instance was destroyed.

Commit:
Adds missing <mutex>, <memory> include files in WebSocket.h

dumganhar commented:
@xiaofeng11 , may i ask when will this Pull Request be merged?

dumganhar commented:
@xiaofeng11 , may i ask when will this Pull Request be merged?
+1

dumganhar commented:
> @dumganhar what is the use case for cocos2d-x ? Is it only valid for cocos2d-js users ?
> BTW, the build is red. If you fix it, I'll merge it.

@ricardoquesada , my changes are also valid for C++ & Lua users.
WebSocket.h/.cpp is a wrapper for libwebsockets, and could be used for C++/Lua/JS users.

deviantmk commented:
bumb

ricardoquesada commented:
@dumganhar the build is still red. could you fix the issues that prevent it from building ok?

dumganhar commented:
@ricardoquesada, 
@xiaofeng11 has been fixing the `red` problem. I don't know the process, so i am asking the status of updating the 3rd `libwebsockets` here. But got no response. :(

pandemosth commented:
@dumganhar could you have a look at my question on https://github.com/cocos2d/cocos2d-x-3rd-party-libs-src/pull/78 regarding why SSL support has been reverted in your change?
Thanks

xiaofeng11 committed:
Merge commit 'refs/pull/14867/head' of https://github.com/cocos2d/cocos2d-x into v3_libwebsockets

xiaofeng11 committed:
libwebsockets have been compiled as static lib, remove libwebsockets.dll from win10 & win8.1 project

xiaofeng11 commented:
after pr https://github.com/cocos2d/cocos2d-x-3rd-party-libs-bin/pull/199 been merged, change deps lib ref in this PR and pull my branch https://github.com/xiaofeng11/cocos2d-x/tree/v3_libwebsockets to combine the change of cocos/2d/win10_props/cocos2d_win10_app.props and cocos/2d/winrt_8.1_props/cocos2d_winrt_8.1_app.props to get pass of jenkins auto build

dumganhar commented:
Hi @pandemosth ,
Don't worry, I think @xiaofeng11 has compiled it with SSL support yet.

xiaofeng11 commented:
@dumganhar The libwebsockets update is compile by cocos2d-x-3rd-party-libs-src compile scripts, it won't have ssl support

pandemosth commented:
Can that be fixed? Websocket client really should support ssl.

> On 14 Feb 2016, at 1:24 PM, xiaofeng11 notifications@github.com wrote:
> 
> @dumganhar The libwebsockets update is compile by cocos2d-x-3rd-party-libs-src compile scripts, it won't have ssl support
> 
> —
> Reply to this email directly or view it on GitHub.

dumganhar commented:
The previous release version of Cocos2dx also doesn't support WebSocket with SSL, right?
Anyway, whether it was supported or not, the changes of this PR isn't relevant to it.
@xiaofeng11 , could you compile `libwebsockets` with SSL support?

xiaofeng11 commented:
@pandamicro @dumganhar Sorry, my mistake. New libwebsocket is compiled with openSSL support.
But I can not sure if Win10 & Win8.1 have same feature, because these lib can not compile with scripts. So I create project file from win32 src and modify code as patch in MS fork to compile these 2 lib

zilongshanren commented:
@dumganhar 
The 3rd party bin has been merged, please check and update the config.json file.
Thanks.

dumganhar commented:
@zilongshanren ,ok, i will.

Commit:
Updates config.json

dumganhar commented:
Update config.json DONE.

xiaofeng11 commented:
@dumganhar as mentioned above, you need pull my branch at https://github.com/xiaofeng11/cocos2d-x/tree/v3_libwebsockets to merge change to cocos/2d/win10_props/cocos2d_win10_app.props and cocos/2d/winrt_8.1_props/cocos2d_winrt_8.1_app.props to get this pr pass jinkens auto build.

dumganhar commented:
@xiaofeng11 ,ok , i will check it.

Commit:
Merge branch 'v3' of git://github.com/cocos2d/cocos2d-x into refactor-websocket

Commit:
Merge branch 'v3_libwebsockets' of https://github.com/xiaofeng11/cocos2d-x into refactor-websocket

dumganhar commented:
@xiaofeng11 , DONE for merging  https://github.com/xiaofeng11/cocos2d-x/tree/v3_libwebsockets

dumganhar commented:
@xiaofeng11 , build passed. Please check.

xiaofeng11 commented:
@pandamicro @zilongshanren I think this pr is OK for merge.

dumganhar commented:
hey, @all, how about this PR now?

dumganhar commented:
Hello, could this Pull Request be merged now ?

zilongshanren commented:
@dumganhar 
I'm fine with this PR, @ricardoquesada  thoughts?

dumganhar commented:
It's really disappointed that there isn't any updates. :(

ricardoquesada commented:
good to me. merged

dumganhar commented:
@ricardoquesada, thank you. :)

stammen commented:
@ricardoquesada @dumganhar This pull request broke the Windows Universal 8.1 ARM builds. What version of libwebsockets did you use? I will compile all of the libs and submit a pull request to cocos2d-x-3rd-party-libs-bin. Should I add OpenSSL support?

zilongshanren commented:
@stammen  maybe @xiaofeng11  knows this issue better, he made all the prebuilt libs for windows universal 8.1.

xiaofeng11 commented:
@stammen This pr use 15c92b1 commit on master branch, which commits on Jan 18, 2016. You can use https://github.com/cocos2d/cocos2d-x-3rd-party-libs-src scripts to get this version.

And could you please build a win32 version either? I have compiled a win32 version, but my version can not compatible with Visual Studio 2015, compile cpp-tests with my version libwebsockets.lib, I get __imp____iob_func & __imp___vsnprintf missing error at link. But it work fine with Visual Studio 2013.

dannyjiajia commented:
hi guys, why disable the `onError` callback in Lua ? https://github.com/cocos2d/cocos2d-x/blob/v3/cocos/scripting/lua-bindings/manual/network/Lua_web_socket.cpp#L136. @dumganhar

roed commented:
When I try the new websockets code, my win32 machine doesn't connect to my node.js backend anymore. It always ends up with ErrorCode::CONNECTION_FAILURE.

dumganhar commented:
@dannyjiajia , Sorry for my late reply, I don't know why it was commented. 
And this pull request didn't comment that line. Anyway, I will find out why it was commented.

dumganhar commented:
@roed , how to reproduce that problem? Could you send me a node.js demo for reproduce? Thanks.

dannyjiajia commented:
@dumganhar thanks, any update?

dumganhar commented:
@dannyjiajia, I checked the git commit tree. Found it was commented at the beginning of v3.x.
Yep, it should not be commented. I created an issue (https://github.com/cocos2d/cocos2d-x/issues/15904) and will try to fix it.

dannyjiajia commented:
@dumganhar thanks :)

dumganhar commented:
@dannyjiajia , the issue was fixed at https://github.com/cocos2d/cocos2d-x/pull/15922
Thanks for your feedback.

roed commented:
@dumganhar I ran the cocos2d-x test case and it did work with a socket.io test server. In my project, I only replaced some libraries and the websocket code, which didn't work. Will do a full upgrade to 3.11.1 and see if the code still breaks. So you can ignore my question, sorry :)

dumganhar commented:
@roed, OK. Feel free of reporting anything after you upgrade to 3.11.1. Thanks.

roed commented:
@dumganhar I upgraded to 3.11.1, win32 works fine but now I have an issue on iOS. My socket.io server code is nothing more than something like:

```
var app = require('http').createServer();
var io = require('socket.io').listen(app);
app.listen(3000);
io.on('connection', function (socket) {
    console.log('Socket connected: ', socket.id);
});
```

Somehow lws_client_connect in websockets.cpp returns a nullptr and doesn't give me anything else in the logging output. Do you have any idea what this could be? I also tried compiling the library myself, but with the same issues.
